src/arcaflow_plugin_sdk/atp.py

```diff
@@ -247,49 +319,53 @@ def read_hello(self) -> HelloMessage:
         message = self.decoder.decode()
         return _HELLO_MESSAGE_SCHEMA.unserialize(message)
 
-    def start_work(self, step_id: str, input_data: any):
+    def start_work(self, run_id: str, step_id: str, config: any):
         """
         After the Hello message has been read, this function starts work in a plugin with the specified data.
         """
-        self.encoder.encode(
+        self.send_runtime_message(
+            MessageType.WORK_START,
+            run_id,
             {
                 "id": step_id,
-                "config": input_data,
+                "config": config,
             }
         )
-        self.stdin.flush()
 
-    def send_signal(self, step_id: str, signal_id: str, input_data: any):
+    def send_signal(self, run_id: str, signal_id: str, input_data: any):
         """
         This function sends any signals to the plugin.
         """
-        self.send_runtime_message(MessageType.SIGNAL, {
-                "step_id": step_id,
+        self.send_runtime_message(
+            MessageType.SIGNAL,
+            run_id,
+        {
                 "signal_id": signal_id,
                 "data": input_data,
             }
         )
 
     def send_client_done(self):
-        self.send_runtime_message(MessageType.CLIENT_DONE, {})
+        self.send_runtime_message(MessageType.CLIENT_DONE, "", "")
 
-    def send_runtime_message(self, message_type: MessageType, data: any):
+    def send_runtime_message(self, message_type: MessageType, run_id: str, data: any):
         self.encoder.encode(
             {
-                "id": message_type.value,
+                "id": message_type,
+                "run_id": run_id,
                 "data": data,
             }
         )
-        self.stdin.flush()
+        self.to_server_pipe.flush()
 
-    def read_results(self) -> (str, any, str):
+    def read_single_result(self) -> (str, str, any, str):
```

Comment:
Tuples are problematic from a maintenance perspective.  Consider using a [`namedtuple`](https://docs.python.org/3/library/collections.html#namedtuple-factory-function-for-tuples-with-named-fields) (or [`NamedTuple`](https://docs.python.org/3.6/library/typing.html?highlight=namedtuple#typing.NamedTuple)) or defining a [`@dataclass`](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass) instead, which allow you to name the members, which frees you from having to get the ordering right and allows you to add context to the values.